src/main/kotlin/org/jitsi/jicofo/jigasi/JigasiDetector.kt

```diff
@@ -0,0 +1,117 @@
+/*
+ * Jicofo, the Jitsi Conference Focus.
+ *
+ * Copyright @ 2018 - present 8x8, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jitsi.jicofo.jigasi
+
+import org.jitsi.impl.protocol.xmpp.XmppProvider
+import org.jitsi.jicofo.JicofoConfig.Companion.config
+import org.jitsi.jicofo.xmpp.BaseBrewery
+import org.jitsi.utils.logging2.createLogger
+import org.jitsi.xmpp.extensions.colibri.ColibriStatsExtension
+import org.json.simple.JSONObject
+import org.jxmpp.jid.EntityBareJid
+import org.jxmpp.jid.Jid
+
+/**
+ * [JigasiDetector] manages the pool of Jigasi instances by joining a "brewery" room that Jigasis connect to
+ * and publish their status in MUC presence.
+ * @author Damian Minkov
+ */
+class JigasiDetector(
+    xmppProvider: XmppProvider,
+    breweryJid: EntityBareJid
+) : BaseBrewery<ColibriStatsExtension>(
+    xmppProvider,
+    breweryJid,
+    ColibriStatsExtension.ELEMENT_NAME,
+    ColibriStatsExtension.NAMESPACE,
+    createLogger().apply { addContext("type", "jigasi") }
+) {
+    override fun onInstanceStatusChanged(jid: Jid, status: ColibriStatsExtension) {}
+    override fun notifyInstanceOffline(jid: Jid) {}
+
+    /**
+     * Selects a jigasi instance which supports transcription.
+     * @param exclude a list of excluded instances.
+     * @param preferredRegions a list of preferred regions.
+     * @return the JID of the selected instance, or  `null` if there are no available jigasis that satisfy the
+     * constraints (are not excluded and support transcription).
+     */
+    fun selectTranscriber(exclude: List<Jid>, preferredRegions: Collection<String>): Jid? =
+        selectJigasi(instances, exclude, preferredRegions, config.localRegion, transcriber = true)
+
+    /**
+     * Selects a jigasi instance which supports SIP.
+     * @param exclude a list of excluded instances.
+     * @param preferredRegions a list of preferred regions.
+     * @return the JID of the selected instance, or  `null` if there are no available jigasis that satisfy the
+     * constraints (are not excluded and support SIP).
+     */
+    fun selectJigasi(exclude: List<Jid>, preferredRegions: Collection<String>): Jid? =
```

Comment:
nit: looks like it would be nice to rename this to `selectSipJigasi` or something